build_tools/cmake/iree_cc_test.cmake

```diff
@@ -107,19 +107,50 @@ function(iree_cc_test)
   set_property(TARGET ${_NAME} PROPERTY DIRECT_DEPS ${_RULE_DEPS})
 
   string(REPLACE "::" "/" _PACKAGE_PATH ${_PACKAGE_NS})
-  set(_NAME_PATH "${_PACKAGE_PATH}:${_RULE_NAME}")
-  add_test(
-    NAME
-      ${_NAME_PATH}
-    COMMAND
-      # We run all our tests through a custom test runner to allow temp
-      # directory cleanup upon test completion.
-      "${CMAKE_SOURCE_DIR}/build_tools/cmake/run_test.${IREE_HOST_SCRIPT_EXT}"
-      "$<TARGET_FILE:${_NAME}>"
-    WORKING_DIRECTORY
-      "${CMAKE_BINARY_DIR}"
+
+  # Case for cross-compiling towards Android.
+  if(ANDROID)
+    set(_ANDROID_REL_DIR "${_PACKAGE_PATH}/${_RULE_NAME}")
+    set(_ANDROID_ABS_DIR "/data/local/tmp/${_ANDROID_REL_DIR}")
+
+    # Define a custom target for pushing and running the test on Android device.
+    set(_RUN_NAME "run_${_PACKAGE_NAME}_${_RULE_NAME}_on_android_device")
+    add_test(
+      NAME
+        ${_RUN_NAME}
+      COMMAND
+        "${CMAKE_SOURCE_DIR}/build_tools/cmake/run_android_test.${IREE_HOST_SCRIPT_EXT}"
+        "${_ANDROID_REL_DIR}/${_NAME}"
     )
+    # Use environment variables to instruct the script to push artifacts
+    # onto the Android device before running the test. This needs to match
+    # with the expectation of the run_android_test.{sh|bat|ps1} script.
+    set(
+      _ENVIRONMENT_VARS
+        TEST_ANDROID_ABS_DIR=${_ANDROID_ABS_DIR}
+        TEST_ARTIFACT=$<TARGET_FILE:${_NAME}>
+        TEST_ARTIFACT_IS_EXECUTABLE=1
```

Comment:
Rather than a singular TEST_ARTIFACT, what about accepting a list of EXECUTABLE and a list of DATA artifacts? If you're going to future proof with making TEST_ARTIFACT_IS_EXECUTABLE configurable :-D

Or if you don't want to cross that bridge yet (passing lists between cmake and bash/bat/powershell sounds not fun), I'd just have a single TEST_EXECUTABLE argument. At the point where we're maybe not setting TEST_ARTIFACT_IS_EXECUTABLE we almost certainly are going to want to accompanying data (e.g. the check tests will need this).

Reply:
Let's handle that when we really need to push multiple files or executables. Testing `.sh`/`.bat`/`.ps1` across Windows/Linux/Android is not fun. ;) 

Yeah, check tests will need data modules, but at the same time it also need the executable. All tests need the same executable actually one way is just to push the executable once using a separate target. Coming in the next patch.